stor/dx.py

```diff
@@ -0,0 +1,513 @@
+        """
+        proj_id = self.canonical_project
+        proj_name = self.virtual_project
+        ans_list = []
+        kwargs = {
+            'only': only,
+            'describe': {'fields': {'name': True, 'folder': True}}
+        }
+        if self._is_folder():
+            kwargs.update({'folder': self.resource})
+        elif self.resource:
+            return ans_list
+        try:
+            obj_dict = dxpy.DXProject(dxid=proj_id).list_folder(**kwargs)
+        except dxpy.exceptions.ResourceNotFound:
+            raise NotFoundError('The specified folder ({}) was not found'.format(
+                self.resource))
+        for key, values in obj_dict.items():
+            for entry in values:
+                if canonicalize:
+                    ans_list.append(DXCanonicalPath('dx://{}:/{}'.format(
+                        proj_id, (entry.lstrip('/') if key == 'folders' else entry['id']))))
+                else:
+                    if key == 'folders':
+                        ans_list.append(DXVirtualPath(self.drive + proj_name + ':' + entry))
+                    else:
+                        ans_list.append(DXVirtualPath(self.drive + proj_name + ':' +
+                                                      entry['describe']['folder'])
+                                        / entry['describe']['name'])
+        return ans_list
+
+    def listdir_iter(self, canonicalize=False):
+        """Iterate the path as a dir, returning top-level directories and files.
+
+        Args:
+            canonicalize (boolean): whether to return canonicalized paths
+
+        Returns:
+            Iter[DXPath]: Iterates over listed files directly within the resource
+        """
+        folders = self.listdir(only='folders', canonicalize=canonicalize)
+        for folder in folders:
+            yield folder
+        for data in self.walkfiles(
+                        canonicalize=canonicalize,
+                        recurse=False
+                    ):
+            yield data
+
+    def walkfiles(self,
+                  pattern=None,
+                  canonicalize=False,
+                  recurse=True,
+                  starts_with=None,
+                  limit=None,
+                  category=None):
+        """Iterates over listed files that match an optional pattern.
+
+        Args:
+            pattern (str): glob pattern to match the filenames against.
+            canonicalize (boolean): whether to return canonicalized paths
+            recurse (boolean): whether to look in subfolders of folder as well
+            starts_with (str): Allows for an additional search path to
+                be appended to the resource of the dx path. Note that this
+                resource path is treated as a directory
+            limit (int): Limit the amount of results returned
+            category (str): Restricting class : One of 'record', 'file', 'gtable,
```

Comment:
`category (str): if specified, only return results of the given class`

why pick the name `category` instead of `classname` here?

Reply:
similar point about how we don't really need to expose a parameter until it's clear there's a need for it.